### PR TITLE
feat: convert to TS and modernize project

### LINKED ISSUE #33: script to generate docs does not work on node 12
```
yarn run v1.9.4
$ babel-node ./scripts/generate_documentation
internal/modules/cjs/loader.js:613
    throw err;
    ^

Error: Cannot find module './array'
Require stack:
- scripts/generate_documentation/find_exports_for_module.js
- scripts/generate_documentation/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:610:15)
    at Function.resolve (internal/modules/cjs/helpers.js:21:19)
    at loadAstForModule (find_exports_for_module.js:13:34)
    at resolveImportDeclaration (find_exports_for_module.js:27:19)
    at find_exports_for_module.js:70:41
    at arrayMap (node_modules/lodash/lodash.js:639:23)
    at Function.map (node_modules/lodash/lodash.js:9556:14)
    at interceptor (node_modules/lodash/lodash.js:16993:35)
    at thru (node_modules/lodash/lodash.js:8797:14)
    at node_modules/lodash/lodash.js:4374:28
error Command failed with exit code 1.
```

### PR DESCRIPTION
**Pull request by ndhoule**
# Features

- Make fake data reproducible by using a seeded data generator
- Add `fake.unique`, which works the same as `fake.array` but generates unique values (or throws if it's unable to)
- `fake.tzid` accepts optional list of TZID strings

# Bugfixes

- `fake.sample` throws when passed array of length `0` to fix possible type mismatch
- `fake.date` validates inputs (`from`/`to`)

# Refactors

- Modernize dependencies / build harness
- Replace test runner (Mocha) with Jest
- Convert project to TypeScript
- Remove Lodash dependency
- Upgrade all dependencies
- Automatically publish new versions from CI

# Removals

- Remove `fake.factory` (no one ever ended up using this)
- Remove `fake.globallyUniqueString` in favor of `fake.unique` and `_.uniqueId` (I can 
be convinced to bring this back, but AFAICT `catalog-api` is its only consumer)
- Remove `inventoryLot`
- Remove `warehouseLocation`
- Unfortunately, I ran out of time to autogenerate docs from the source. I'm looking into `typedoc` as a means to accomplish this, but I think this PR represents enough of an improvement that it seems worth going forth with this. For now, the API docs in the README should be updated manually. (Closes #33.)

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import fake from '.';
+
+describe('the default export', function() {
+  // TODO: split these into smaller tests
+  it('works', function() {
```
Comment (serhalp) [inline src/index.test.ts]: :face_vomiting:

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,14 @@
+# UNRELEASED
+
+- feat: use Chance.js to generate all fake data
+- chore: convert project from Flow to TypeScript
+- chore: replace test runner (Mocha) with Jest
+- chore: use `@goodeggs/toolkit` to lint project
+- remove `globallyUniqueString`
```
[the remark this answers is not in the document]
Comment (leaopedro) [reply CHANGELOG.md]: There are actually some files in `catalog-api` using `factory()`: see [this](https://github.com/goodeggs/catalog-api/search?q=%22fake.factory%28%22&unscoped_q=%22fake.factory%28%22), please advise on how to replace it.

File: package.json [hunk 3]
```diff
@@ -2,15 +2,19 @@
   "name": "fake-eggs",
   "version": "4.7.1",
   "description": "Generate Good Eggs-flavored data for development / test fixtures",
-  "main": "index.js",
-  "types": "./index.d.ts",
+  "main": "dist/index.js",
   "scripts": {
-    "build": "npm run docs && babel src -d . && flow-copy-source src .",
+    "build": "yarn run clean && yarn run build:src && yarn run build:types && yarn run build:flow_types",
+    "build:src": "babel --extensions .es6,.js,.ts,.es,.jsx,.tsx,.mjs --source-maps --ignore 'src/**/*.test.ts' --out-dir dist/ src/",
```
[the remark this answers is not in the document]
Comment (ndhoule) [reply package.json]: We don't need to--this is the built-in Babel extension list with ts/tsx extensions added in.

File: package.json [hunk 4]
```diff
@@ -2,15 +2,19 @@
   "name": "fake-eggs",
   "version": "4.7.1",
   "description": "Generate Good Eggs-flavored data for development / test fixtures",
-  "main": "index.js",
-  "types": "./index.d.ts",
+  "main": "dist/index.js",
```
[the remark this answers is not in the document]
Comment (ndhoule) [reply package.json]: or `build/`? :man_shrugging: we're pretty inconsistent.

My impression is that `lib` does have a meaning that is not "built source", but if someone could just tell me exactly what the hell is supposed to go in `lib`, then I guess I could be on board? I don't really care that much.

**Comment by shermam** [comment]
Hey @serhalp do you think we could merge this? 
I would really like to use the `unique` function in Garbanzo.

**Comment by serhalp** [comment]
@shermam I'll try to get to this soon

**Comment by shermam** [comment]
> @shermam I'll try to get to this soon

Thanks a lot!

**Comment by serhalp** [comment]
@shermam published as fake-eggs@5.0.0!

**Pull request merged**